### PR TITLE
 denc: slightly optimize container_base::bound_encode. 

### PR DESCRIPTION
**Pull request by rzarzynski**
This cosmetics is just a side quest of poking with `bufferlist`. Sending as clean-up.

### CODE REVIEW

File: src/include/denc.h [hunk 1]
```diff
@@ -882,7 +882,12 @@ namespace _denc {
     static void bound_encode(const container& s, size_t& p, uint64_t f = 0) {
       p += sizeof(uint32_t);
       if constexpr (traits::bounded) {
-        if (!s.empty()) {
+        const auto elem_num = s.size();
+        // intensionally not calling container's empty() method to not prohibit
+        // compiler from optimizing the check if it and the ::size() operate on
+        // different memory (observed when std::list::empty() works on pointers,
+        // not the size field).
```
Comment (tchaikov) [inline src/include/denc.h]: @rzarzynski i am wondering if you are optimizing the case where

1. `C<Ts...>::size()` is O(1)
1. if it's not O(1), the majority `std::list` we are encoding is not empty.

but the `std::list::size()` GCC-7 offered by SCLs is O(n). so i can hardly tell if this change is a pure win.

**Comment by tchaikov** [comment]
@rzarzynski could you please drop bcbe38a from this PR so we can discuss it separately?

File: src/include/denc.h (hunk 1 above)
Comment (rzarzynski) [reply src/include/denc.h]: Yeah, this change targets solely the 1st case. For C++11-incompliant implementations of `std::list::empty` a penalty can be expected. The question is for how long we need to support those cases. For sure backporting the commit isn't an option but maybe the next release is free of the legacy?

Comment (tchaikov) [reply src/include/denc.h]: GCC-8 from SCLs is still using the C++11-incompliant ABI.

Comment (tchaikov) [reply src/include/denc.h]: if we are able to detect this at build time, probably we can conditionalize on this. the ugliest change i can think of is to add a cmake option to do this. for instance, pass `-DCXX11_COMPATIBLE=OFF` only on centos/rhel7.

i can help on this if you think it's a viable way.

Comment (rzarzynski) [reply src/include/denc.h]: Thanks for the commit, @tchaikov!

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-08-21_12:58:47-rados-wip-kefu-testing-2019-08-21-1445-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-23_00:12:22-rados-wip-kefu-testing-2019-08-21-1445-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-08-23_03:12:13-rados-wip-kefu-testing-2019-08-21-1445-distro-basic-smithi/

**Pull request merged**